Separate third pillar mandate e-mail into two parts

MatsJohanson opened the pull request:
In the first e-mail, we're sending out payment details for the third pillar and giving some tips on how to DCA your contributions over time.

The second e-mail will be sent 3 days later to users who have not joined our second pillar, reminding them why our fund is a good option.

ErkoRisthein commented on src/main/java/ee/tuleva/onboarding/mandate/email/MandateEmailService.java:
```diff
@@ -21,57 +25,90 @@
+
+  private void sendThirdPillarPaymentDetailsEmail(
+      User user, Mandate mandate, UserPreferences contactDetails, Locale locale) {
+    String subject =
```
Why hardcoded translations?

ErkoRisthein commented on src/main/java/ee/tuleva/onboarding/mandate/email/MandateEmailService.java:
```diff
@@ -21,57 +25,90 @@
+    String subject =
+        locale == DEFAULT_LOCALE
+            ? "Vaata oma teine sammas üle!"
+            : "Check where your second pillar is invested!";
```
Why hardcoded translations? Use the `MessageSource` bean instead

ErkoRisthein commented on src/main/java/ee/tuleva/onboarding/mandate/email/MandateEmailSender.java:
```diff
@@ -18,22 +18,20 @@
-    emailService.sendMandate(
-        event.getUser(), event.getMandate(), pillarSuggestion, contactDetails, event.getLocale());
+    return new PillarSuggestion(
+        event instanceof ThirdPillarAfterMandateSignedEvent ? 2 : 3,
```
I avoid using `instanceof` at all costs. Usually, it means that I'm breaking the **polymorphism** principle of Object Oriented Programming.

Any way to solve this polymorphically instead? i.e. `event.getPillar()`?

ErkoRisthein replied:
I would maybe try to get rid of the `SecondPillarAfterMandateSignedEvent` and `ThirdPillarAfterMandateSignedEvent` subclasses and add a `getPillar()` method directly to `AfterMandateSignedEvent`.

ErkoRisthein commented on src/main/java/ee/tuleva/onboarding/mandate/email/MandateEmailService.java:
```diff
@@ -21,57 +25,90 @@
+      default -> throw new IllegalArgumentException(
+          "Unknown pillar: " + pillarSuggestion.getOtherPillar());
+    }
+  }
```
Any way to solve this polymorphically instead of branching? We should avoid logic like 
`if 2nd pillar then do this, else if 3rd pillar then do that, else throw exception` at _all costs_. Ideally, we would do it in 1 place only and then use polymorphism in the rest of the codebase.

ErkoRisthein commented on src/test/groovy/ee/tuleva/onboarding/mandate/email/MandateEmailServiceSpec.groovy:
```diff
@@ -3,83 +3,159 @@ package ee.tuleva.onboarding.mandate.email
+    1 * emailService.send(user, message)
+  }
+
+  def "Send third pillar suggest second pillar email"() {
```
User `@Unroll` for Data Driven Tests

MatsJohanson replied:
From Spock 2, all tests are unrolled by default so less boilerplate!

ErkoRisthein commented on src/test/groovy/ee/tuleva/onboarding/mandate/email/MandateEmailServiceSpec.groovy:
```diff
@@ -3,83 +3,159 @@ package ee.tuleva.onboarding.mandate.email
-        true                | true                        | true     || ["mandate", "pillar_2"]
-    }
-
-    @Unroll
```
Why remove `@Unroll`?

ErkoRisthein commented:
If you see any way to improve the design of the code, then always do that as well, as part of the task itself. I.e. following the [boy scouts rule](https://stackoverflow.com/a/37314918/2102820).
Let's have our codebase gradually improving, not degrading.

Btw, what's the code coverage of this pull request? We used to have automated checks with codecov.io, but we removed it due to security issues. Maybe it makes sense to revisit this (but let's make sure we don't execute any remote bash scripts - rather commit the script to the repo).

MatsJohanson commented:
> If you see any way to improve the design of the code, then always do that as well, as part of the task itself. I.e. following the [boy scouts rule](https://stackoverflow.com/a/37314918/2102820).
> Let's have our codebase gradually improving, not degrading.
> 
> Btw, what's the code coverage of this pull request? We used to have automated checks with codecov.io, but we removed it due to security issues. Maybe it makes sense to revisit this (but let's make sure we don't execute any remote bash scripts - rather commit the script to the repo).

Did the fixes in a separate PR here: https://github.com/TulevaEE/onboarding-service/pull/412

I didn't go all the way with polymorphism fixes, but at least removed the unneeded `else` edge cases, since they are already checked on the PillarSuggestion level. Will improve the code further in the future.

I added a ticket to turn on CodeCoverage again.
